index.rst

```diff
@@ -1,57 +1,451 @@
+
+While the FITS header keys listing the extensions could previously be hard-coded into :class:`~lsst.afw.image.ExposureInfo`'s depersistence code, the depersistence code for generic components must search the header for keys of the expected format.
+Queries for ``[arbitrary string]_ID`` are vulnerable to false positives from header keys like ``VISIT_ID`` or ``CCD_ID``; if associated values are small integers, then there is no way to  distinguish such keys from real archive IDs.
+
+We therefore added a second convention for archive IDs, of the form ``ARCHIVE_ID_[component]``.
+Old components are depersisted using the ``*_ID`` syntax, to retain compatibility with old files, while new ones are depersisted using ``ARCHIVE_ID_*``.
+The new code strips both sets of header keywords when they are detected; new files read using old code may have leftover ``ARCHIVE_ID_`` keywords.
+
+.. _newpersistence:
+
+Alternative Persistence Frameworks
+==================================
+
+Rationale
+---------
+
+Despite recent improvements to :class:`lsst.afw.image.ExposureInfo`, our ability to attach data to an :class:`~lsst.afw.image.Exposure` is limited by the ``lsst.afw.table.io`` persistence framework.
+We still have circular dependencies within ``afw``, we still need persistable objects to be written in C++, and we still need to work around the non-optimal nature of ``lsst.afw.table.io`` itself.
+To go further, we need to replace ``lsst.afw.table.io`` with a well-maintained package that, unlike ``afw.table``, was designed for persistence.
+
+Adoption Constraints
+--------------------
+
+We are looking for a persistence framework that meets the following criteria:
+
+* it must have a GPL3-compatible license.
+* it must allow persistence of both C++ and Python objects.
+* it must allow persistence to multiple formats, including FITS, JSON, and YAML.
+  Compatibility with JSON and YAML implies that the framework must be able to represent objects as key-value pairs.
+* it must allow versioning of persistence formats.
+* it must allow depersistence of old files written with the ``lsst.afw.table.io`` framework.
+* it must allow efficient storage of arrays, particularly images, but not necessarily in formats other than FITS.
+* it must correctly depersist polymorphic types that are stored in C++ by their base class (e.g., :class:`lsst.afw.detection.Psf`), reproducing their exact type (e.g., :class:`lsst.meas.algorithms.ImagePsf`).
+* it must be able to read in part of a persisted object, such as only the WCS from a persisted :class:`~lsst.afw.image.Exposure`.
+* it should be able to store relationships between objects that refer to each other.
+  This would allow us to separately store composite objects and their components, such as the individual PSFs used to create a :class:`lsst.meas.algorithms.CoaddPsf`, simplifying provenance tracking.
+* it should persist files in a human-readable form, where practical, as a debugging aid.
```

Comment:
> I did wonder about the “adoption constraints” for an alternative persistence framework. Some of these are obviously essential, but others are less so (the ability to persist to YAML, for instance, and arguably depersistence of old files). My understanding is that many of the original goals for this work are stymied until we solve the persistence issue, so we should think very critically about what's actually necessary here. Adding a few sentences to the document on this would be helpful.

These were mostly copied from my notes from February, with a few added in (e.g., versioning) based on @TallJimbo's intro. @TallJimbo, can you address @jdswinbank's question about how necessary some of these "musts" are?

Reply:
I agree with the general thought that we're not going to be able to satisfy all of these, now that we've taken a look at the landscape.  But I don't think I have any special insight into which ones are nonnegotiable and which aren't; it's all a big box of tradeoffs.  Maybe just soften the overall language above the list?